src/scanner.c

```diff
@@ -0,0 +1,205 @@
+#include <assert.h>
+#include <stdio.h>
+#include <string.h>
+#include <tree_sitter/parser.h>
+
+#define MAX(a, b) ((a) > (b) ? (a) : (b))
+
+enum TokenType {
+    HERE_STRING_BODY,
+};
+
+typedef struct {
+    uint32_t len;
+    uint32_t cap;
+    char *data;
+} String;
+
+static String string_new() {
+    return (String){.cap = 16, .len = 0, .data = calloc(1, sizeof(char) * 17)};
+}
+
+static void string_resize(String *vec, uint32_t cap) {
+    void *block = realloc(vec->data, (cap + 1) * sizeof(vec->data[0]));
+    assert(block != NULL);
+    vec->data = block;
+    memset(
+        vec->data + vec->len, 0, ((cap + 1) - vec->len) * sizeof(vec->data[0]));
+    vec->cap = cap;
+}
+
+static void string_grow(String *vec, uint32_t cap) {
+    if (vec->cap < cap) {
+        string_resize(vec, cap);
+    }
+}
+
+static void string_push(String *vec, int32_t elem) {
+    if (vec->cap == vec->len) {
+        string_resize(vec, MAX(16, vec->len * 2));
+    }
+    vec->data[vec->len++] = (char)elem;
```

Comment:
there is a bug that `int32_t` character is converted to `char`.
`int32_t` can hold unicode character, while `char` only hold ASCII character.
I suggest push 4 times for each char in `int32_t`.